kong/clustering/services/sync/rpc.lua

```diff
@@ -0,0 +1,339 @@
+    local ok, err = kong.db.clustering_data_planes:upsert({ id = node_id }, {
+      last_seen = ngx.time(),
+      hostname = node_id,
+      ip = kong.rpc:get_peer_ip(node_id),   -- try to get the correct ip
+      version = "3.8.0.0",    -- XXX TODO: get from rpc call
+      sync_status = CLUSTERING_SYNC_STATUS.NORMAL,
+      config_hash = fmt("%032d", default_namespace_version),
+      rpc_capabilities = rpc_peers and rpc_peers[node_id] or {},
+    }, { ttl = purge_delay })
+    if not ok then
+      ngx_log(ngx_ERR, "unable to update clustering data plane status: ", err)
+    end
+
+    -- is the node empty? If so, just do a full sync to bring it up to date faster
+    if default_namespace_version == 0 or
+       self.strategy:get_latest_version() - default_namespace_version > FULL_SYNC_THRESHOLD
+    then
+      -- we need to full sync because holes are found
+
+      ngx_log(ngx_INFO,
+              "[kong.sync.v2] database is empty or too far behind for node_id: ", node_id,
+              ", current_version: ", default_namespace_version,
+              ", forcing a full sync")
+
+
+      local deltas, err = declarative.export_config_sync()
+      if not deltas then
+        return nil, err
+      end
+
+      -- wipe dp lmdb, full sync
+      return gen_delta_result(deltas, true)
+    end
+
+    local res, err = self.strategy:get_delta(default_namespace_version)
+    if not res then
+      return nil, err
+    end
+
+    if #res == 0 then
+      ngx_log(ngx_DEBUG,
+              "[kong.sync.v2] no delta for node_id: ", node_id,
+              ", current_version: ", default_namespace_version,
+              ", node is already up to date" )
+      return gen_delta_result(res, false)
+    end
+
+    -- some deltas are returned, are they contiguous?
+    if res[1].version == default_namespace.version + 1 then
+      -- doesn't wipe dp lmdb, incremental sync
+      return gen_delta_result(res, false)
+    end
+
+    -- we need to full sync because holes are found
+    -- in the delta, meaning the oldest version is no longer
+    -- available
+
+    ngx_log(ngx_INFO,
+            "[kong.sync.v2] delta for node_id no longer available: ", node_id,
+            ", current_version: ", default_namespace_version,
+            ", forcing a full sync")
+
+    local deltas, err = declarative.export_config_sync()
+    if not deltas then
+      return nil, err
+    end
+
+    -- wipe dp lmdb, full sync
+    return gen_delta_result(deltas, true)
+  end)
+end
+
+
+function _M:init_dp(manager)
+  -- DP
+  -- Method: kong.sync.v2.notify_new_version
+  -- Params: new_versions: list of namespaces and their new versions, like:
+  -- { { namespace = "default", new_version = 1000, }, }
+  manager.callbacks:register("kong.sync.v2.notify_new_version", function(node_id, new_versions)
+    -- TODO: currently only default is supported, and anything else is ignored
+    for namespace, new_version in pairs(new_versions) do
```

Comment:
it seems that it could be simplified and access the default namespace with: `new_versions.default` like above.

Reply:
Yes, it is, I will do it, thanks.